pkg/virt-controller/watch/vm.go

```diff
@@ -2899,6 +2899,75 @@ func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
 	vm.Status.VolumeRequests = tmpVolRequests
 }
 
+func filterLiveUpdatableVolumes(lastSeenVMSpec *virtv1.VirtualMachineSpec, vm *virtv1.VirtualMachine) {
+	oldVols := make(map[string]virtv1.Volume)
+	for _, v := range lastSeenVMSpec.Template.Spec.Volumes {
+		// The old volumes which can be updated are only hotpluggable
+		// volumes, such as PVCs and DVs.
+		if v.DataVolume != nil || v.PersistentVolumeClaim != nil {
+			oldVols[v.Name] = v
+		}
+	}
+
+	// Evaluate if any volume has changed or has been added
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		oldVol, ok := oldVols[v.Name]
+		switch {
+		// Changes for hotlpugged volumes are valid
+		case storagetypes.IsHotplugVolume(&v):
+			delete(oldVols, v.Name)
+		// The volume has been freshly added
+		case !ok:
+			return
+		// The volume has changed
+		case !equality.Semantic.DeepEqual(oldVol, v):
+			return
+		default:
+			delete(oldVols, v.Name)
+		}
+	}
+	// Evaluate if any volumes were removed and they were hotplugged volumes
+	for _, v := range oldVols {
+		if !storagetypes.IsHotplugVolume(&v) {
+			return
+		}
+	}
+
+	// Evaluate the disks. They need to remain equal except for the ones
+	// associated to hotplugged volumes
+	oldDisks := storagetypes.GetDisksFromVolumes(&lastSeenVMSpec.Template.Spec)
+	// Remove the disks associated to an hotpluggable volume
+	for _, v := range lastSeenVMSpec.Template.Spec.Volumes {
+		if storagetypes.IsHotplugVolume(&v) {
+			delete(oldDisks, v.Name)
+		}
+	}
+	newDisks := storagetypes.GetDisksFromVolumes(&vm.Spec.Template.Spec)
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		oldDisk, okOld := oldDisks[v.Name]
+		newDisk := newDisks[v.Name]
```

Comment:
I think it's possible to have a volume without a disk and if so `newDisk` will be nil

Reply:
Really? What is this case? How can the VM use the volume if there isn't a disk?

Reply:
Ah it might be a filesystem